implementations/rust/ockam/ockam_api/tests/utils/mod.rs

```diff
@@ -0,0 +1,375 @@
+                                )
+                                .await
+                            });
+                        }
+                        Disruption::DropPacketsAfter(drop_packets_after) => {
+                            tokio::spawn(async move {
+                                relay_stream_drop_packets(
+                                    incoming_read,
+                                    outgoing_write,
+                                    drop_packets_after,
+                                )
+                                .await
+                            });
+                            tokio::spawn(async move {
+                                relay_stream_drop_packets(
+                                    outgoing_read,
+                                    incoming_write,
+                                    drop_packets_after,
+                                )
+                                .await
+                            });
+                        }
+                        Disruption::PacketsOutOfOrderAfter(packet_out_of_order_after) => {
+                            tokio::spawn(async move {
+                                relay_stream_packets_out_of_order(
+                                    incoming_read,
+                                    outgoing_write,
+                                    packet_out_of_order_after,
+                                )
+                                .await
+                            });
+                            tokio::spawn(async move {
+                                relay_stream_packets_out_of_order(
+                                    outgoing_read,
+                                    incoming_write,
+                                    packet_out_of_order_after,
+                                )
+                                .await
+                            });
+                        }
+                    }
+                });
+            }
+        });
+    }
+
+    PassthroughServerHandle {
+        chosen_addr,
+        destination,
+        close,
+    }
+}
+
+async fn relay_stream_limit_bandwidth(
+    mut read_half: OwnedReadHalf,
+    mut write_half: OwnedWriteHalf,
+    max_bytes_per_second: Option<usize>,
+) {
+    let mut bytes_counter = 0;
+    let mut buffer = vec![0; 64 * 1024];
+    loop {
+        let read = match read_half.read(&mut buffer).await {
+            // socket closed
+            Ok(0) => return,
+            Ok(n) => n,
+            Err(e) => {
+                error!("Failed to read from socket; err = {:?}", e);
+                return;
+            }
+        };
+
+        if let Err(e) = write_half.write_all(&buffer[0..read]).await {
+            error!("Failed to write to socket; err = {:?}", e);
+            return;
+        }
+
+        bytes_counter += read;
+        if let Some(max_bytes_per_second) = max_bytes_per_second {
+            if bytes_counter >= max_bytes_per_second {
```

Comment:
Does that work correctly if max_bytes_per_second is much smaller than 64 * 1024? And do we need it to work with such values?

Reply:
The limit works, but portals go in timeout since the delay gets too big. 

We shouldn't timeout just because the bandwidth is limited, since this could limit deployment environments. Having a lower limit of 64kb/s is already a small compromise IMO.

Reply:
@davide-baldo let's say the limit is 1024 bytes/second, then let's say you successfully read 64 kbytes to the `buffer`. That means you need to wait for another 64 seconds before the next read, but it looks like this code will only sleep for 1 second instead. Am I missing something?
I see two options:
1. Something like this:
```
while bytes_counter >= max_bytes_per_second {
    bytes_counter -= max_bytes_per_second;
    tokio::time::sleep(Duration::from_secs(1)).await;
}
```
2. Even better: reduce `buffer` size to `max_bytes_per_second`.

I think option 2 is better, but it's also far from perfect and won't behave precisely if traffic load isn't uniform (do we use it in such conditions, or is the load uniformly high?).

Reply:
The second option was the first implementation but it was impractical for high limits.
Another option i'm testing is to sleep based on how many bytes just passed. Kinda reflect real-world scenario where you get an initial peak before the buffers gets full.

Reply:
dynamic sleep based on the max allow seems to be working great :+1: